examples/tutorials/Part 12 bis - Encrypted Training on MNIST.ipynb

```diff
@@ -0,0 +1,664 @@
+    "        if i < n_train_items / args.batch_size\n",
+    "    ]\n",
+    "    \n",
+    "    test_loader = torch.utils.data.DataLoader(\n",
+    "        datasets.MNIST('../data', train=True, download=True, transform=transformation),\n",
+    "        batch_size=args.test_batch_size\n",
+    "    )\n",
+    "    \n",
+    "    private_test_loader = [\n",
+    "        (secret_share(data), secret_share(target.float()))\n",
+    "        for i, (data, target) in enumerate(test_loader)\n",
+    "        if i < n_test_items / args.test_batch_size\n",
+    "    ]\n",
+    "    \n",
+    "    return private_train_loader, private_test_loader\n",
+    "    \n",
+    "    \n",
+    "private_train_loader, private_test_loader = get_private_data_loaders(\n",
+    "    precision_fractional=args.precision_fractional,\n",
+    "    workers=workers,\n",
+    "    crypto_provider=crypto_provider\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Model specification\n",
+    "\n",
+    "Here is the model that we will use, it's a rather simple one but [it has proved to perform reasonably well on MNIST](https://towardsdatascience.com/handwritten-digit-mnist-pytorch-977b5338e627)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class Net(nn.Module):\n",
+    "    def __init__(self):\n",
+    "        super(Net, self).__init__()\n",
+    "        self.fc1 = nn.Linear(28 * 28, 128)\n",
+    "        self.fc2 = nn.Linear(128, 64)\n",
+    "        self.fc3 = nn.Linear(64, 10)\n",
+    "\n",
+    "    def forward(self, x):\n",
+    "        x = x.view(-1, 28 * 28)\n",
+    "        x = F.relu(self.fc1(x))\n",
+    "        x = F.relu(self.fc2(x))\n",
+    "        x = self.fc3(x)\n",
+    "        return x"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Training and testing functions\n",
+    "\n",
+    "The training is done almost as usual, the real difference is that we can't use losses like negative log-likelihood (`F.nll_loss` in PyTorch) because it's quite complicated to reproduce these functions with SMPC. Instead, we use a simpler Mean Square Error loss."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train(args, model, private_train_loader, optimizer, epoch):\n",
+    "    model.train()\n",
+    "    for batch_idx, (data, target) in enumerate(private_train_loader): # <-- now it is a private dataset\n",
+    "        start_time = time.time()\n",
+    "        \n",
+    "        optimizer.zero_grad()\n",
+    "        \n",
+    "        output = model(data)\n",
+    "        \n",
+    "        # loss = F.nll_loss(output, target)  <-- not possible here\n",
+    "        batch_size = output.shape[0]\n",
+    "        loss = ((output - target)**2).sum().refresh()/batch_size\n",
```

Comment:
Why do you need to use `.refresh()` here? Can't it be hidden?

Reply:
I don't see an easy way of doing it currently, unfortunately :/

Reply:
What does refresh() do?

Reply:
So... Division by integers C (or truncation) is a real pain with additive shared tensor:
For fixed precision, values x live in Zq so if x < q/2 you just do x/C and if x > q/2 you do (x - q) / C + q.
But now if you use additive shared tensors, now you have x1 and x2 in Zq, not observable at the same time, with x = x1 + x2 allegedly in Zq. When dividing by C you can't know if x > q/2 and no worker with one share can also decide this...
So we implement a rule of thumb of a paper which works in most cases, provided that the shares are fresh: ie x1 and x2 look like random values in Zq. That's why I need to **refresh the shares by adding shares of 0** because otherwise the division fails.

To the question: should be automatize refreshing under the hood? I really don't know, I need to better understand when it's needed and when it's not, etc.

Anyway thanks for spotting this, I agree it needs further explanations.

Reply:
I vote always refresh() - better to be slow and not scary for new users than the other way around :)

Reply:
New User: "WTF! Why doesn't division work?!?!?!" 

is better than

New User: Hmm - this encryption thing is kindof slow - but i guess it's worth it and will get faster as technology advances.

Reply:
Yes, but for some mysterious reason if I refresh everytime it doesn't work neither! I need to find when refreshing should automatically be performed.

Reply:
Sounds good - no need to block for that feature as it doesn't pertain to this tutorial directly.